### PR TITLE
fix: use focus listeners for message read status

### PR DESCRIPTION
**Pull request by jeremy-croff**
### Details
Implement the correct isVisible and isFocus methods for the electron Visibility Util.
When this util was originally implemented, electron did not have win.isVisible yet. original implementation for reference:
https://github.com/Expensify/App/commit/e7150f92bfd50e5d46ba65e85163bdac808a2229

### Fixed Issues

$ https://github.com/Expensify/App/issues/34829
PROPOSAL: https://github.com/Expensify/App/issues/34829#issuecomment-1904919422

### Tests

QA steps;

Given two logged in users, atleast one the electron app
1. Open Desktop App
2. Login as User A
3. Open Web App/Device
4. Login as User B

Given created chat
5. Click the green plus icon in the bottom left 
6. press Start Chat
7. type the email from person you want to chat with
8. hit add to group to create the chat

Scenario receives notifications and marks as read when clicking on it

When chat is not focused
9. From User A select a chat of Concierge or another one that's not user B
10. From User B sent a message to User A
11. User A received Notification, wait until it docks
12. Open Notification center and click on this received notification
13. Verify the chat is marked as read

When chat is focused and desktop app is minimized
14. From User A select a chat of user B
15. Minimize Desktop App of User A
15. From User B sent a message to User A
16. User A received Notification, wait until it docks
17. Open Notification center and click on this received notification
18. Verify the chat is marked as read

When chat is not focused and desktop app is already visibile
19. Open Desktop App of User A to front view ( visible state on monitor )
20. From User A select a chat of Concierge or another one that's not user B
21. From User B sent a message to User A
22. User A received Notification, wait until it docks
23. Open Notification center and click on this received notification
24. Verify the chat is marked as read

Scenario: chat doesn't receive notifications, but chat marks as read

When: chat is focused and desktop app is visibile
25. Open Desktop App of User A to front view ( visible state on monitor )
26. From User A select a chat of user B
27. From User B sent a message to User A
28. User A doesnt receive notification
29. Verify the chat is marked as read

When: chat is focused and desktop app is occluded but not minimized
30. Open Desktop App of User A to front view ( visible state on monitor )
31. From User Aselect a chat of user B
32. From User B sent a message to User A
33. User A doesnt receive notification
34. Verify the chat is marked as read

- [x] Verify that no errors appear in the JS console

### Offline tests

1. Disconnect the internet on user B
2. Send messages from User A to B
3. Connect the internet on User B
4. Verify notifications come through

### QA Steps

Given the 
Scenario: two logged in users, atleast one the electron app
and
Scenario: created chat between the users
( or see steps above on how to create)

When user A is on another chat like Concierge, or has the electron app minimized:
-Send a message from User B to user A
-User A received Notification
- Wait until Notification goes to the Notification center dock
-Open Notification center and click on this received notification
- Verify the chat is marked as read

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/157416545/886488dc-7a96-44d0-a188-6a6edfd2c5ab

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/157416545/8d815595-742d-4cd0-81c4-839ff846b1e5

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/157416545/3af91579-5926-4481-bb2a-d3c464761efa

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/157416545/087bc3d9-6334-41f6-a935-b83847d059ca

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/157416545/cb89c836-fd88-4adb-bdd4-1d296c631a4b

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/157416545/ca690b5f-2b17-43ff-9526-341b7c26fea7

</details>

### CODE REVIEW

**Comment by jeremy-croff** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by jeremy-croff** [comment]
recheck

**Comment by jeremy-croff** [comment]
I want to make a special note on this test case I ran:
```
Scenario: report is focused and desktop app is occluded but not minimized
When
- Open Desktop App of User A to front view ( visible state on monitor )
- From User B select a chat of user A
- From User B sent a message to User A

Then
- User A doesnt receive notification
- Verify the chat is marked as read

```

When the screen is occluded, but the chat is open. The customer does not receive an notification. This seems to be a recent bug with electron visibility https://github.com/electron/electron/issues/38955 on certain devices.

This is a regression because the isFocus would have still sent a notification.

**Comment by jeremy-croff** [comment]
recheck

**Comment by Julesssss** [comment]
I was [incorrectly](https://github.com/Expensify/App/issues/34829#issuecomment-1906834450) assigned here. Reassigning to @blimpich, and manually assigning @ishpaul777 for the C+ review

**Comment by jeremy-croff** [comment]
> @jeremy-croff Can you please record a video scenario explained, with and without the additional changes
> 
> For future reference: please make sure to add full link to issue, before opening PR for review so automation can work properly
> 
> ```
> Fixed Issues
> $ https://github.com/Expensify/App/issues/<issueid>
> ```

Thank you for all the tips! Will be sure to do so.

Before:
1. bug with visibility and occlusion on electron ( regression )
2. bug with mark existing open chat as read (existing)


https://github.com/Expensify/App/assets/157416545/4bf02447-6c15-49c4-bf4a-dfd40e280bf2


After
1. fixes both  bugs

https://github.com/Expensify/App/assets/157416545/a88c09aa-a4d0-48c7-8b62-0134d5cefca7

**Comment by jeremy-croff** [comment]
> Also can you please update test Steps with a numbered list with expected result on each step (if any), see any of the other PRs like: #34560, #34563 etc.

updated

**Comment by jeremy-croff** [comment]
> Thanks! Bug 1, seems like a bug caused by the changes with the PR, will so should be fixed Please push a fix. Bug 2, If this is a existing bug and will be fixed with additional changes, please note that there is no additional compensation for this and comes with the risk of causing a regression which will cause a "-50%" if found during regression period

Because the regression is caused by the usage of isVisible per https://github.com/electron/electron/issues/38955 with no known workarounds. Reverting back to isFocus undoes my proposal.

In order to leverage isFocus as a solution, the onfocus listener must be utilized and Bug 2 comes free out of the box!

I think my hands are tied here.

I have merged in the additional proposed changed.

File: src/pages/home/report/ReportActionsList.js [hunk 1]
```diff
@@ -202,18 +203,23 @@ function ReportActionsList({
...
+    const readNewestReportActionOnFocus = () => {
+        if(!ReportUtils.isUnread(report)){
+            return
+        }
+        if (Visibility.hasFocus() && scrollingVerticalOffset.current < MSG_VISIBLE_THRESHOLD) {
+            Report.readNewestAction(report.reportID);
+        } else {
+            readActionSkipped.current = true;
+        }
+    }
```
Comment (blimpich) [inline src/pages/home/report/ReportActionsList.js:215]: Can we rewrite this like so?

```diff
diff --git a/src/pages/home/report/ReportActionsList.js b/src/pages/home/report/ReportActionsList.js
index 05f73032d7..b2ea93da28 100644
@@ -204,10 +204,7 @@ function ReportActionsList({
     }, [report.reportID]);
 
     const readNewestReportActionOnFocus = () => {
-        if(!ReportUtils.isUnread(report)){
-            return
-        }
-        if (Visibility.hasFocus() && scrollingVerticalOffset.current < MSG_VISIBLE_THRESHOLD) {
+        if (ReportUtils.isUnread(report) && Visibility.hasFocus() && scrollingVerticalOffset.current < MSG_VISIBLE_THRESHOLD) {
             Report.readNewestAction(report.reportID);
         } else {
             readActionSkipped.current = true;

```

**Review by blimpich — CHANGES_REQUESTED** [review]

File: src/pages/home/report/ReportActionsList.js (hunk 1 above)
Comment (jeremy-croff) [reply src/pages/home/report/ReportActionsList.js]: Makes sense, updated.
Thanks!

**Comment by blimpich** [comment]
@ishpaul777 Yeah ideally we should fix the bug in Electron since it is the root cause. It doesn't look like there is much movement there though so having this sit here seems like not a great option either. I just posted [a comment](https://github.com/electron/electron/issues/38955#issuecomment-1908825876) there asking a maintainer who seemed close to fixing it whether they will end up fixing it anytime soon. Lets see if they respond. If we don't think that it'll get fixed upstream anytime soon then I think we should try to make this PR work

**Comment by jeremy-croff** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by jeremy-croff** [comment]
recheck

**Comment by jeremy-croff** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by jeremy-croff** [comment]
recheck

**Comment by jeremy-croff** [comment]
> Thanks for your thoughts! @blimpich, Lets put this for hold for response and revisit the issue next week.

Moved the lo

**Comment by jeremy-croff** [comment]
> Hey @jeremy-croff, Changes seems to be causing regressions Steps:
> 
> 1. start a chat between account A and account B
> 2. with account A, open any other chat, for example, Concierge Chat, open any other window
> 3. Send message to account A from B
> 4. manually focus on to Desktop App (not from notification)
> 
> expected result, chat with B is unread Actual result, chat with B is marked as read
> 
> And Honestly I am not confident with changes applied, the ideal solution is to hold for a upstream fix for [electron/electron#38955](https://github.com/electron/electron/issues/38955) then revisit this issue instead of workaroud, @blimpich any thoughts ?

I've moved the logic that only reads the active report to be also used in the onfocus hook to address this. Good catch.

I'm still pending a full regression and can do so to have this ready while waiting on clarification if we should go with an upstream fix.

A case to make for not going with an upstream fix is because I believe visibility is not the right trigger for marking things as read. I currently believe focus is.

**Comment by blimpich** [comment]
So far no response from upstream repo. I just emailed the engineer in case the GH comment didn't get through to them. I say we give that a day or two and if no response then lets go ahead with this.

File: desktop/main.js [hunk 2]
```diff
@@ -557,6 +557,11 @@ const mainWindow = () => {
 
                 ipcMain.on(ELECTRON_EVENTS.REQUEST_VISIBILITY, (event) => {
                     // This is how synchronous messages work in Electron
+                    // eslint-disable-next-line no-param-reassign
+                    event.returnValue = browserWindow && !browserWindow.isDestroyed() && browserWindow.isVisible();
+                });
+
+                ipcMain.on(ELECTRON_EVENTS.REQUEST_HAS_FOCUS, (event) => {
```
[the remark this answers is not in the document]
Comment (jeremy-croff) [reply desktop/main.js:564]: I believe this change is desired, this implements the correct focus logic for electron. Before it defaulted to true.

[the remark this answers is not in the document]
Comment (jeremy-croff) [reply desktop/main.js]: @ishpaul777 Updated PR. thanks for the suggestion. 
It's indeed not necessary for this solution. The initial thought that moving other logic over to focus would be more consistent in cases where windows are occluded and are still marking it as read. But it's out of this issues scope.

**Comment by iwiznia** [comment]
What's the deal with this PR? More than a month without an update and this is blocking https://github.com/Expensify/App/issues/35085

The issue says [this](https://github.com/Expensify/App/issues/34829#issuecomment-1980478783) but I think that PR has been already deployed.

**Comment by blimpich** [comment]
@iwiznia this issue is logistically messy. We originally were waiting for an upstream maintainer of Electron to patch Electron, and eventually they did but then the fix got reverted, then we also thought #33680 might fix this issue's root cause. 

@ishpaul777 it looks like #33680 got deployed to production, can we retest this to see if it reproduces and then if not, move forward with this PR? I've lost faith that the upstream maintainers of Electron are going to fix this for us.

**Pull request closed without merging**